L3-205: including cancellation examples for making and viewing. (#22)
digicatapult/dscp-documentation: docs/l3/api_flow.md

  }
```

Note: At any time after a `match2` is proposed, and before it reaches `acceptedFinal` state, any of its members can reject the `match2` using `POST /v1/match2/{id}/rejection`. Once a `match2` is rejected, it is permanently closed.

> 

Reviewer: ```suggestion

When a `match2` is in a state of `acceptedFinal` it can be cancelled by either Member A or Member B, if seen fit i.e. order or capacity is no longer available for a specific reason.

1. For example, cancelling `match2` as Member A using `POST /v1/match2/{match2Id}]/cancellation`:

___remove___```js
{
    "id": "1d007432-519b-41ca-8d36-ded5b3d40d52",
    "state": "submitted",
    "localId": "fef4a34b-e3a2-46bb-b4a0-c67cc32eac94",
    "apiType": "match2",
    "transactionType": "cancellation",
    "submittedAt": "2023-09-27T15:13:18.464Z",
    "updatedAt": "2023-09-27T15:13:18.464Z"
}
___remove__```

```

> ## Cancelling match2
> 
> 1. Viewing all `match2` cancellation transactions can be performed by `memberA` or `memberB` by making **GET** request to `/v1/match2/{match2Id}/cancellation` endpoint. Takes a `match2id` as a parameter property. Match2 must be in `finalAcceptance` state.

Reviewer: ```suggestion
## Cancelling match

When a `match2` is in a state of `acceptedFinal` it can be cancelled by either Member A or Member B, if seen fit i.e. order or capacity is no longer available for a specific reason.

1. For example, cancelling `match2` as Member A using `POST /v1/match2/{match2Id}]/cancellation`:

*remove*```js
{
    "id": "1d007432-519b-41ca-8d36-ded5b3d40d52",
    "state": "submitted",
    "localId": "fef4a34b-e3a2-46bb-b4a0-c67cc32eac94",
    "apiType": "match2",
    "transactionType": "cancellation",
    "submittedAt": "2023-09-27T15:13:18.464Z",
    "updatedAt": "2023-09-27T15:13:18.464Z"
}
*remove*```

2. Viewing all `match2` cancellation transactions can be performed by `memberA` or `memberB` by making **GET** request to `/v1/match2/{match2Id}/cancellation` endpoint. Takes a `match2id` as a parameter property. Match2 must be in `finalAcceptance` state.
```

#### response example
```js
[
  ...,
  {
    "id": "a3acc8b9-1e7b-47de-bb91-fdcb81c41ed7",
    "state": "inBlock",
    "localId": "56261a42-e046-488a-875d-7ed9e93d0000",
    "apiType": "match2",
    "transactionType": "cancellation",
    "submittedAt": "2023-09-27T10:34:17.030Z",
    "updatedAt": "2023-09-27T10:34:18.043Z"
  },
  {
    "id": "3ba8e8b9-1e7b-47de-bb91-fdcb81c41ed7",
    "state": "inBlock",
    "localId": "56261a42-e046-488a-875d-7ed9e93dde09",
    "apiType": "match2",
    "transactionType": "cancellation",
    "submittedAt": "2023-09-26T10:34:17.030Z",
    "updatedAt": "2023-09-26T10:34:18.043Z"
  }
]

/* Response headers
 access-control-allow-origin: * 
 connection: keep-alive 
 content-length: 246 
 content-type: application/json; charset=utf-8 
 date: Tue,26 Sep 2023 10:37:13 GMT 
 etag: W/"f6-jordvLkfdT0QftNEAa6voZtlS6o" 
 keep-alive: timeout=5 
 x-powered-by: Express 
 */
```

> 2. In order to retrieve a specific transaction for a specific `match2`. Make a call to the `/v1/match2/{match2Id}/cancellation/{cancellationId}` endpoint containing both params e.g. `cancellationId` and `match2Id`
> 
> #### response example
> ```js
> {
>   "id": "3ba8e8b9-1e7b-47de-bb91-fdcb81c41ed7",
>   "state": "inBlock",
>   "localId": "56261a42-e046-488a-875d-7ed9e93dde09",
>   "apiType": "match2",
>   "transactionType": "cancellation",
>   "submittedAt": "2023-09-26T10:34:17.030Z",
>   "updatedAt": "2023-09-26T10:34:18.043Z"
> }
> /* Response headers
>  access-control-allow-origin: * 
>  connection: keep-alive 
>  content-length: 244 
>  content-type: application/json; charset=utf-8 
>  date: Tue,26 Sep 2023 10:38:10 GMT 
>  etag: W/"f4-i3WHp+uSae6+YYElA/k0xWqaguI" 
>  keep-alive: timeout=5 
>  x-powered-by: Express 
>  */
> ```
> 3. Cancels an existing `match2` that is currently in `finalAccepted` state. It must require to have an attachmentId in the request body
> 
> #### request body example
> ```sh
> curl -X 'POST' \
>   '/v1/match2/{match2id}/cancellation' \
>   -H 'accept: application/json' \
>   -H 'Content-Type: application/json' \
>   -d '{
>     "attachmentId": "a789ad47-91c3-446e-90f9-a7c9b233eaf8"
>   }'
> ```
> 
> #### response example
> ```js
> {
>   "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
>   "state": "submitted",
>   "apiType": "match2",
>   "transactionType": "creation",
>   "submittedAt": "2023-09-26T12:06:55.035Z",
>   "updatedAt": "2023-09-26T12:06:55.035Z"
> }
> ```
> 
> # Generic
> > Not Found - 404 
> ```js
> {
>   "name": "string",
>   "message": "string",
>   "stack": "string",
>   "code": 0,
>   "item": "string"
> }
> ```

Reviewer: ```suggestion
## Cancelling match2
When a `match2` is in a state of `acceptedFinal` it can be cancelled by either Member A or Member B, if seen fit i.e. order or capacity is no longer available for a specific reason. It is mandatory that a comment is attached to detail a reason for cancellation.

1. Comment file uploaded in `POST /v1/attachment` as Member A's comment:

*remove*```
The strawberry's have gone bad!
*remove*```

2. For example, cancelling `match2` as Member A using `POST /v1/match2/{match2Id}]/cancellation`:

*remove*```js
{
    "id": "8fb8aef6-0944-4367-bce0-fbbcde48ac54",
    "state": "submitted",
    "localId": "6ad909c6-c06c-463d-bbed-4cdec8781a86",
    "apiType": "match2",
    "transactionType": "cancellation",
    "submittedAt": "2023-09-29T12:49:38.458Z",
    "updatedAt": "2023-09-29T12:49:38.458Z"
}
*remove*```

3. Viewing a `match2` cancellation transaction can be performed by Member A or Member B by using `GET /v1/match2/{match2Id}/cancellation`:

*remove*```js
    {
        "id": "8fb8aef6-0944-4367-bce0-fbbcde48ac54",
        "state": "finalised",
        "localId": "6ad909c6-c06c-463d-bbed-4cdec8781a86",
        "apiType": "match2",
        "transactionType": "cancellation",
        "submittedAt": "2023-09-29T12:49:38.458Z",
        "updatedAt": "2023-09-29T12:49:55.772Z"
    }
*remove*```

4. In order to retrieve a specific transaction for a specific `match2`, use `GET /v1/match2/{match2Id}/cancellation/{cancellationId}`:

*remove*```js
{
    "id": "8fb8aef6-0944-4367-bce0-fbbcde48ac54",
    "state": "finalised",
    "localId": "6ad909c6-c06c-463d-bbed-4cdec8781a86",
    "apiType": "match2",
    "transactionType": "cancellation",
    "submittedAt": "2023-09-29T12:49:38.458Z",
    "updatedAt": "2023-09-29T12:49:55.772Z"
}
*remove*```
```

  }
```

Note: At any time after a `match2` is proposed, and before it reaches `acceptedFinal` state, any of its members can reject the `match2` using `POST /v1/match2/{id}/rejection`. Once a `match2` is rejected, it is permanently closed.

## Cancelling match2
When a `match2` is in a state of `acceptedFinal` it can be cancelled by either Member A or Member B, if seen fit i.e. order or capacity is no longer available for a specific reason. It is mandatory that a comment is attached to detail a reason for cancellation.

1. Comment file uploaded in `POST /v1/attachment` as Member A's comment:

```

> The strawberry's have gone bad!

Reviewer 2: plural of strawberry is strawberries

Reviewer: Thanks Doctor!

  }
```

Note: At any time after a `match2` is proposed, and before it reaches `acceptedFinal` state, any of its members can reject the `match2` using `POST /v1/match2/{id}/rejection`. Once a `match2` is rejected, it is permanently closed.

## Cancelling match2

> 1. Viewing all `match2` cancellation transactions can be performed by `memberA` or `memberB` by making **GET** request to `/v1/match2/{match2Id}/cancellation` endpoint. Takes a `match2id` as a parameter property. Match2 must be in `finalAcceptance` state.

Reviewer 3: Surely this will only let you view a single cancellation transaction rather than ALL match2 cancellation transactions?

Author: there can be scenario where it could be multiple for a same match2 @mattdean-digicatapult might be able to provide more context. I think due to race condition but memory is a bit rusty

Reviewer: I'm going to suggest adding a step for actually cancelling the match2 which is missing from the doc, in addition to just viewing a cancelled match.

#### response example
```js
[
  ...,
  {
    "id": "3ba8e8b9-1e7b-47de-bb91-fdcb81c41ed7",
    "state": "inBlock",
    "localId": "56261a42-e046-488a-875d-7ed9e93dde09",
    "apiType": "match2",
    "transactionType": "cancellation",
    "submittedAt": "2023-09-26T10:34:17.030Z",
    "updatedAt": "2023-09-26T10:34:18.043Z"

>   }

Reviewer 3: If you are going to show that it is a list of items then you should show 2 cancellations as an example, instead of just 1.

Author: left a `...` and also it's an array... thought it's self explanatory. Added one more record.

]

/* Response headers
 access-control-allow-origin: * 
 connection: keep-alive 
 content-length: 246 
 content-type: application/json; charset=utf-8 
 date: Tue,26 Sep 2023 10:37:13 GMT 
 etag: W/"f6-jordvLkfdT0QftNEAa6voZtlS6o" 
 keep-alive: timeout=5 
 x-powered-by: Express 
 */
```

2. In order to retrieve a specific transaction for a specific `match2`. Make a call to the `/v1/match2/{match2Id}/cancellation/{cancellationId}` endpoint containing both params e.g. `cancellationId` and `match2Id`

#### response example
```js
{
  "id": "3ba8e8b9-1e7b-47de-bb91-fdcb81c41ed7",
  "state": "inBlock",
  "localId": "56261a42-e046-488a-875d-7ed9e93dde09",
  "apiType": "match2",
  "transactionType": "cancellation",
  "submittedAt": "2023-09-26T10:34:17.030Z",
  "updatedAt": "2023-09-26T10:34:18.043Z"
}
/* Response headers
 access-control-allow-origin: * 
 connection: keep-alive 
 content-length: 244 
 content-type: application/json; charset=utf-8 
 date: Tue,26 Sep 2023 10:38:10 GMT 
 etag: W/"f4-i3WHp+uSae6+YYElA/k0xWqaguI" 
 keep-alive: timeout=5 
 x-powered-by: Express 
 */
```

> 3. Cancels an existing `match2` that is currently in `finalAccepte` state. It must require to have an attachmentId in the request body

Reviewer 3: ```suggestion
3. Cancels an existing `match2` that is currently in `finalAccepted` state. It must require to have an attachmentId in the request body
```
